DQMServices/Core/interface/MonitorElement.h

```diff
@@ -15,58 +23,140 @@
+    MonitorElementData::Value const &value;
+  };
+  // TODO: can this be the same type, just const?
+  struct AccessMut {
+    std::unique_lock<dqmmutex> guard_;
+    MonitorElementData::Key const &key;
+    MonitorElementData::Value &value;
+  };
+
+  struct MutableMonitorElementData {
+    MonitorElementData data_;
+    dqmmutex lock_;
+    Access access() { return Access{std::unique_lock<dqmmutex>(lock_), data_.key_, data_.value_}; }
+    AccessMut accessMut() { return AccessMut{std::unique_lock<dqmmutex>(lock_), data_.key_, data_.value_}; }
+  };
+
   /** The base class for all MonitorElements (ME) */
   class MonitorElement {
-    friend class DQMStore;
+    friend dqm::dqmstoreimpl::DQMStore;
     friend DQMService;
 
   public:
-    struct Scalar {
-      int64_t num;
-      double real;
-      std::string str;
-    };
+    using Scalar = MonitorElementData::Scalar;
+    using Kind = MonitorElementData::Kind;
+
+  protected:
+    DQMNet::CoreObject data_;  //< Core object information.
+    // TODO: we only use the ::Value part so far.
+    // Still using the full thing to remain compatible with the new ME implementation.
+
+    std::atomic<MonitorElementData const *> frozen_;    // only set if this ME is in a product already
+    std::atomic<MutableMonitorElementData *> mutable_;  // only set if there is a mutable copy of this ME
+    /** 
+     * To do anything to the MEs data, one needs to obtain an access object.
+     * This object will contain the lock guard if one is needed. We differentiate
+     * access for reading and access for mutation (denoted by `Access` or
+     * `const Access`, however, also read-only access may need to take a lock
+     * if it is to a mutable object. Obtaining mutable access may involve
+     * creating a clone of the backing data. In this case, the pointers are
+     * updated using atomic operations. It can happen that reads go to the old
+     * object while a new object exists already, but this is fine; concurrent
+     * reads and writes can happen in arbitrary order. However, we need to
+     * protect against the case where clones happen concurrently and avoid
+     * leaking memory or loosing updates in this case, using atomics.
+     * We want all of this inlined and redundant operations any copies/refs
+     * optimized away.
+     */
+    const Access access() const {
+      // First, check if there is a mutable object
+      auto mut = mutable_.load();
+      if (mut) {
+        // if there is a mutable object, that is the truth, and we take a lock.
+        return mut->access();
+      }  // else
+      auto frozen = frozen_.load();
+      if (frozen) {
+        // in case of an immutable object read from edm products, create an
+        // access object without lock.
+        return Access{std::unique_lock<dqmmutex>(), frozen->key_, frozen->value_};
+      }
+      // else
+      assert(!"Attempting to access MonitorElement not backed by any data!");
```

Comment:
Make this an exception instead of an assert and the compiler warning will go away. You are getting the warning since the compiler knows assert can be turned off with CPP macros.

Reply:
Ah yes. I saw that one but forgot.